iwyu.cc

```diff
@@ -1392,12 +1379,23 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     if (!IsInMacro(use_loc))
       return use_loc;
 
-    // If we're in a macro, and the macro file forward-declares decl, make sure
-    // we keep that forward-declaration.
-    const NamedDecl* fwd_decl = GetForwardDeclareInSameFile(decl, use_loc);
+    // If the macro definition file contains a forward-declaration for decl,
+    // we treat that as a use-attribution hint. See below for details.
+    const FileEntry* macro_def_file = GetFileEntry(use_loc);
+    const NamedDecl* fwd_decl = nullptr;
+    for (const NamedDecl* redecl : GetClassRedecls(decl)) {
+      if (GetLocation(redecl).isFileID() &&
+          GetFileEntry(redecl) == macro_def_file &&
+          IsForwardDecl(redecl)) {
+        fwd_decl = redecl;
+        break;
+      }
+    }
```

Comment:
What case `isFileID` is covering? Just curious, nothing wrong with it.

It might be useful to mention that for macros it's not required for forward declaration to be before usage. Not sure about it, I struggled with it at first but your response in PR made it clear.

And the biggest problem for me was how did we get from `use_loc` to `macro_def_file` because use should be not in macro definition file. And the answer to that is that [`GetFileEntry` looks past spelling location](https://github.com/include-what-you-use/include-what-you-use/blob/master/iwyu_location_util.h#L145-L168) and arrives at macro definition file. I don't know, having something like
`const FileEntry* macro_def_file = GetFileEntry(GetSpellingLoc(use_loc));`
might express the intention better.

Anyway, how forward declare hint is supposed to work with macro token concatenation? Because I came up with test case
```c++
// tests/cxx/issue350-i1.h
class debugLogger {
public:
  void Log(const char* msg) {}
};

// ================
// tests/cxx/issue350-d1.h
#include "tests/cxx/issue350-i1.h"

#define LOG(level, x) \
  level ## Logger().Log(x);

// ================
// tests/cxx/issue350.cpp
#include "tests/cxx/issue350-d1.h"

void foo() {
    LOG(debug, "test");
}
```

And IWYU recommends to `#include "tests/cxx/issue350-i1.h"` for `debugLogger`. Though my understanding is that it shouldn't because there is no forward declaration and `issue350-d1.h` is responsible for including `issue350-i1.h`.